aten/src/ATen/cuda/tunable/GemmCublasLt.h

```diff
@@ -0,0 +1,595 @@
+    }
+    at::cuda::blas::CuBlasLtMatrixLayout Ddesc(c_dtype, params->m, params->n, params->ldc);
+
+    int batch_size = GetBatchFromParams<T>(params);
+    if (batch_size > 1) {
+        int64_t stride_a = GetStrideAFromParams<T>(params);
+        int64_t stride_b = GetStrideBFromParams<T>(params);
+        int64_t stride_c = GetStrideCFromParams<T>(params);
+        Adesc.setAttribute(CUBLASLT_MATRIX_LAYOUT_BATCH_COUNT, batch_size);
+        Bdesc.setAttribute(CUBLASLT_MATRIX_LAYOUT_BATCH_COUNT, batch_size);
+        Cdesc.setAttribute(CUBLASLT_MATRIX_LAYOUT_BATCH_COUNT, batch_size);
+        Adesc.setAttribute(CUBLASLT_MATRIX_LAYOUT_STRIDED_BATCH_OFFSET, stride_a);
+        Bdesc.setAttribute(CUBLASLT_MATRIX_LAYOUT_STRIDED_BATCH_OFFSET, stride_b);
+        Cdesc.setAttribute(CUBLASLT_MATRIX_LAYOUT_STRIDED_BATCH_OFFSET, stride_c);
+    }           
+
+    at::cuda::blas::CuBlasLtMatmulPreference preference;
+    // See https://github.com/pytorch/pytorch/issues/73328 for reasoning behind
+    // setting this to 1M.
+    size_t workspaceSize = at::cuda::blas::_getWorkspaceSize();
+    preference.setAttribute(CUBLASLT_MATMUL_PREF_MAX_WORKSPACE_BYTES, workspaceSize);
+
+    uint32_t a_alignment = at::cuda::blas::_getAlignment(reinterpret_cast<uintptr_t>(params->a));
+    uint32_t b_alignment = at::cuda::blas::_getAlignment(reinterpret_cast<uintptr_t>(params->b));
+    uint32_t c_alignment = at::cuda::blas::_getAlignment(reinterpret_cast<uintptr_t>(params->c));
+    preference.setAttribute(CUBLASLT_MATMUL_PREF_MIN_ALIGNMENT_A_BYTES, a_alignment);
+    preference.setAttribute(CUBLASLT_MATMUL_PREF_MIN_ALIGNMENT_B_BYTES, b_alignment);
+    preference.setAttribute(CUBLASLT_MATMUL_PREF_MIN_ALIGNMENT_C_BYTES, c_alignment);
+
+    auto& allocator = *::c10::cuda::CUDACachingAllocator::get();
+    auto workspace = allocator.allocate(workspaceSize);
+    TORCH_CHECK(workspace.get() != nullptr, "OOM trying to allocate workspace for cublaslt");
+
+    cublasLtMatmulHeuristicResult_t heuristicResults[8];
+    int returnedResults = 0;
+
+    TORCH_CUDABLAS_CHECK(cublasLtMatmulAlgoGetHeuristic(
+        ltHandle,
+        computeDesc.descriptor(),
+        Adesc.descriptor(),
+        Bdesc.descriptor(),
+        Cdesc.descriptor(),
+        Ddesc.descriptor(),
+        preference.descriptor(),
+        8,
+        heuristicResults,
+        &returnedResults));
+    
+    if (returnedResults == 0) {
+        TORCH_CUDABLAS_CHECK(CUBLAS_STATUS_NOT_SUPPORTED);
+    }
+
+    std::vector<std::pair<std::string, std::unique_ptr<Callable<ParamsT>>>> ret;
+
+    for (int i = 0; i < returnedResults; i++) {
+        auto algo = heuristicResults[i].algo;
+        auto callable = std::make_unique<CublasltGemmOp<T, ALayout, BLayout, ParamsT>>(algo);
+        std::string type_string = c10::str("Gemm_Cublaslt_", _charFromcublasOp(transa_outer), _charFromcublasOp(transb_outer), "_", i);
```

Comment:
I'm assuming there's a second level of key-ing that includes information like the concrete shapes, alignment(s), etc?

Reply:
yep, that's done through (https://github.com/bilal2vec/pytorch/blob/5690e16e0ea55e9cc1ee492be2162d7e85b72b96/aten/src/ATen/cuda/tunable/TunableOp.h#L74-L82) and is caches the heuristic to use in a .csv file for later use